### PR TITLE
WIP: Adds notes for discussion about reparametrization and loss computation

### PR DESCRIPTION
**Pull request by itellaetxe**
PR for discussion on the code.

The exponential terms in the computations might explode, should we clip them? Also, KL loss tends to negative values the further the training goes. Strange.

### CODE REVIEW

**Comment by itellaetxe** [comment]
@pjsjongsung @jhlegarreta 

Even after clipping expontential operations to (1e-10 and 1e10), I am observing that the outputs of the decoder (`z_mean`, `z_log_var`) are very high, resulting in extremely high values after sampling with the reparametrization trick. This results in super high MSE loss and the model just does not converge anywhere. I am not incorporating the `kl_loss` term to the total loss yet until I solve this numerical instability issue.

I want to explore how Batch Normalization could help because it prevents the values from going nuts, normalizing its inputs to mean=0 and variance=1. I will add a comment after checking this.

**Comment by itellaetxe** [comment]
> I want to explore how Batch Normalization could help because it prevents the values from going nuts, normalizing its inputs to mean=0 and variance=1. I will add a comment after checking this.

After adding Batch Normalization to the Encoder, I trained the model for a few epochs and saw that the reconstruction loss (MSE) is going down to not-crazy values (<100). Batchnorm was added after each convolutional layer, but not to the dense layers that output `z_mean` and `z_log_var`, to not enforce super strict mean=0 and var=1 in their outputs, as I suspect this may constrain too much the latent space.

Then, I added the KL loss term to the total loss (`total_loss = reconstruction_loss + kl_loss`) and I also saw that the model can be trained. Good news!

Now I need to see whether adding Batch Normalization to the Decoder will benefit convergence.


For now the results look like this (For `kl_loss: 30.6345 - total loss: 54.8003 - reconstruction_loss: 24.1658`). It kind of looks like the FiberCup!
image

**Review by jhlegarreta — COMMENTED** [review]
OK, so following https://github.com/itellaetxe/tractoencoder_gsoc/pull/1#issuecomment-2225337790 looks like batch norm and the clipping worked?

So the comments about using `ops.exp` are now redundant?

The input data stayed the same (no normalization), right?

**Comment by itellaetxe** [comment]
@jhlegarreta 
> OK, so following [#1 (comment)](https://github.com/itellaetxe/tractoencoder_gsoc/pull/1#issuecomment-2225337790) looks like batch norm and the clipping worked?

Yes, at least for training it worked. I am positive that it will also work for inference because for sampling from the latent space and decoding no batch normalization is used. If the latent space is correctly regularized during training (e.g.: no big holes in the manifold), it should be fine.

> So the comments about using `ops.exp` are now redundant?

Yes, I will update them.

> The input data stayed the same (no normalization), right?

Yes, I didn't normalize the input data, but this could be another strategy. I should investigate which is the most convenient from both.

**Comment by jhlegarreta** [comment]
OK.

> Yes, I didn't normalize the input data, but this could be another strategy. I should investigate which is the most convenient from both.

OK, was just asking because it was unclear from the comments. In my early experiments this did not help for the regular AE. But yes, we see that the translation to a VAE is requiring additional thought. For now, I would leave the model train for e.g. >100 epochs with the same input that we were using with the AE to see whether the reconstruction is good enough, and then assess if that was enough.

**Comment by itellaetxe** [comment]
> OK, was just asking because it was unclear from the comments. In my early experiments this did not help for the regular AE. But yes, we see that the translation to a VAE is requiring additional thought. For now, I would leave the model train for e.g. >100 epochs with the same input that we were using with the AE to see whether the reconstruction is good enough, and then assess if that was enough.

@jhlegarreta @pjsjongsung 
I trained the network for 120 epochs with a learning rate of 0.000334 (half of the original) and a batch size of 128. I got this losses `kl_loss: 11.2033 - loss: 21.6290 - reconstruction_loss: 10.4257` (_loss = kl_loss + reconstruction_loss_). Both plausible and implausible fibers are included in the training data. See the reconstruction of the training data below:
image


Reconstruction on the validation data (plausible fibers):
image

Reconstruction on the test set data (plausible fibers):
image


I am still figuring out how to implement the cyclic annealing for the KL loss (cyclically adjusting the weight of the KL loss component in the total loss sum). WIP

**Comment by jhlegarreta** [comment]
OK, thanks @itellaetxe, these look reasonable results to move forward. Out of curiosity, is the coronal (~depth) being recovered correctly? What happens if you set the weight of the KL loss to 0 (i.e. train as if it were an AE with the batch norm added layers)? What is the reconstruction loss for the AE that we had before and this latter case?

**Comment by pjsjongsung** [comment]
> I am still figuring out how to implement the cyclic annealing for the KL loss (cyclically adjusting the weight of the KL loss component in the total loss sum). WIP

Why do we need the cyclic annealing of the KL loss? We could simply test with different fixed loss weights for now right? The [ beta-VAE paper](https://openreview.net/forum?id=Sy2fzU9gl) might prove to have some insight on what weights you can use?

**Comment by itellaetxe** [comment]
> OK, thanks @itellaetxe, these look reasonable results to move forward. Out of curiosity, is the coronal (~depth) being recovered correctly?

@jhlegarreta 
No, the depth is not recovered correctly, the same "2D" representation is learned. From what I discussed with my colleague Jorge, this may be related to how much we compress the information from the last convolutional layer to the latent vector, suggesting that we may be losing some data properties in the way, because the dimension changes in the data are too big. Perhaps we could try to make this transition smoother by adding an intermediate layer in the future.

> What happens if you set the weight of the KL loss to 0 (i.e. train as if it were an AE with the batch norm added layers)? What is the reconstruction loss for the AE that we had before and this latter case?

I tried to train the model yesterday but got interrupted every time (external actors...). I am training it now. I will update you as soon as I get the results.

UPDATE: The loss goes down to ~5.5 in the no-KL-VAE. Similar to what I got with the vanilla-AE. Training has not finished yet, but I would say it performs comparably.

**Comment by itellaetxe** [comment]
> Why do we need the cyclic annealing of the KL loss? We could simply test with different fixed loss weights for now right? The [ beta-VAE paper](https://openreview.net/forum?id=Sy2fzU9gl) might prove to have some insight on what weights you can use?

@pjsjongsung
I will look into this. I got the idea of implementing the cyclic annealing from [this paper](http://biorxiv.org/lookup/doi/10.1101/2023.02.24.529954). Just before the `C` section, they state this:
image

**Comment by jhlegarreta** [comment]
This is probably good to go as it is, and the experiment about the loss weighting can be done in a separate PR. Same comment concerning the coronal axis reconstruction issue; it may not happen for realistic on real data. We keep in mind the possibilities discussed earlier today.

So feel free to merge @itellaetxe whenever you need it in `main`.

**Comment by jhlegarreta** [comment]
BTW, we can avoid duplicating all the code in 9d71633d66301a7dfb516b77aa9d4da022b82108 just by adding a weighting term to the loss that can be set as an input parameter in the regular training script.

**Comment by itellaetxe** [comment]
@jhlegarreta 

> This is probably good to go as it is, and the experiment about the loss weighting can be done in a separate PR. Same comment concerning the coronal axis reconstruction issue; it may not happen for realistic on real data. We keep in mind the possibilities discussed earlier today.
So feel free to merge @itellaetxe whenever you need it in `main`.

Got it, I will merge it today.

> BTW, we can avoid duplicating all the code in https://github.com/itellaetxe/tractoencoder_gsoc/commit/9d71633d66301a7dfb516b77aa9d4da022b82108 just by adding a weighting term to the loss that can be set as an input parameter in the regular training script.

This is also true, I did it mainly because I was paranoid about Keras going crazy on some internal variables that could affect the training... Out of fear basically 😅. I will remove it because it makes no sense to have it there.

Thanks for the comments.

**Comment by itellaetxe** [comment]
@pjsjongsung @jhlegarreta 
I finished running the no-KL VAE training experiment, and the results are good, I would say. After 120 epochs with a learning rate of 0.000344 (half of the vanilla-AE), the reconstructed input looks like this:

image

The "funny" thing is that the streamlines are so far from the origin that I could not visualize them in the first try (I am using MITK and Surfice), so I thought something had gone wrong in the training. Then I chased them through the visualization space until I found them. The screenshot above is the best view I could get.
I would say that with a learning rate of 0.000688 the results would arrive to the same point as the AE, also with 120 epochs. The learning rate was setto 0.000344 because I have had problems with the VAE in the beginning, so I forgot to set it back to 0.000688.

**Comment by jhlegarreta** [comment]
> The "funny" thing is that the streamlines are so far from the origin

Not sure if I understand this. As if they were translated from where they are originally? Weird, should not be the case, as we train on the MSE as well. Same goes for the scale/size. Can you load both the test data input to the VAE and the reconstructed and see where they land?
